File: apps/frontend/src/services/grpc/analytics.service.ts
/**
 * Analytics gRPC Service
 * Teacher-specific analytics for dashboard, students, and exams
 */

// gRPC-Web imports
import { AnalyticsServiceClient } from '@/generated/v1/AnalyticsServiceClientPb';
import {
  GetTeacherDashboardRequest,
  GetTeacherDashboardResponse,
  GetTeacherStudentsRequest,
  GetTeacherStudentsResponse,
  GetTeacherExamsRequest,
  GetTeacherExamsResponse,
  TeacherDashboardData as PbTeacherDashboardData,
  StudentData as PbStudentData,
  ExamData as PbExamData,
  PerformanceTrend as PbPerformanceTrend,
  ExamPerformance as PbExamPerformance,
} from '@/generated/v1/analytics_pb';
import { RpcError } from 'grpc-web';

// gRPC client utilities
import { GRPC_WEB_HOST, getAuthMetadata } from './client';

// Error handling
function handleGrpcError(error: RpcError): string {
  console.error('gRPC Error:', error);
  switch (error.code) {
    case 3: return error.message || 'Invalid input provided';
    case 5: return 'Data not found';
    case 7: return 'Permission denied';
    case 14: return 'Service temporarily unavailable';
    case 16: return 'Authentication required';
    default: return error.message || 'An unexpected error occurred';
  }
}

// ===== gRPC CLIENT INITIALIZATION =====

// Uses GRPC_WEB_HOST which routes through API proxy (/api/grpc) by default
// ✅ FIX: Add format option to match proto generation config (mode=grpcwebtext)
const analyticsServiceClient = new AnalyticsServiceClient(GRPC_WEB_HOST, null, {
  format: 'text', // Use text format for consistency with proto generation
  withCredentials: false,
  unaryInterceptors: [],
  streamInterceptors: []
});

// ===== TYPE DEFINITIONS =====

export interface TeacherDashboardData {
  totalExams: number;
  totalStudents: number;
  averageScore: number;
  passRate: number;
  activeStudents: number;
  completionRate: number;
  trends: PerformanceTrend[];
  topExams: ExamPerformance[];
}

export interface PerformanceTrend {
  date: string;
  averageScore: number;
  passRate: number;
  attemptCount: number;
}

export interface ExamPerformance {
  examId: string;
  examTitle: string;
  attemptCount: number;
  averageScore: number;
  passRate: number;
}

export interface StudentData {
  userId: string;
  email: string;
  firstName: string;
  lastName: string;
  totalCourses: number;
  totalExamResults: number;
  averageScore: number;
  completionRate: number;
  status: string;
  lastActivity?: Date;
}

export interface ExamData {
  id: string;
  title: string;
  description: string;
  subject: string;
  grade: number;
  difficulty: string;
  examType: string;
  status: string;
  durationMinutes: number;
  totalPoints: number;
  passPercentage: number;
  questionCount: number;
  attemptCount: number;
  averageScore: number;
  createdAt: Date;
  publishedAt?: Date;
}

// ===== TYPE MAPPERS =====

function mapTeacherDashboardFromPb(pbData: PbTeacherDashboardData): TeacherDashboardData {
  const dataObj = pbData.toObject();
  return {
    totalExams: dataObj.totalExams,
    totalStudents: dataObj.totalStudents,
    averageScore: dataObj.averageScore,
    passRate: dataObj.passRate,
    activeStudents: dataObj.activeStudents,
    completionRate: dataObj.completionRate,
    trends: dataObj.trendsList.map(mapPerformanceTrendFromPb),
    topExams: dataObj.topExamsList.map(mapExamPerformanceFromPb),
  };
}

function mapPerformanceTrendFromPb(pbTrend: PbPerformanceTrend.AsObject): PerformanceTrend {
  return {
    date: pbTrend.date,
    averageScore: pbTrend.averageScore,
    passRate: pbTrend.passRate,
    attemptCount: pbTrend.attemptCount,
  };
}

function mapExamPerformanceFromPb(pbExam: PbExamPerformance.AsObject): ExamPerformance {
  return {
    examId: pbExam.examId,
    examTitle: pbExam.examTitle,
    attemptCount: pbExam.attemptCount,
    averageScore: pbExam.averageScore,
    passRate: pbExam.passRate,
  };
}

function mapStudentFromPb(pbStudent: PbStudentData.AsObject): StudentData {
  return {
    userId: pbStudent.userId,
    email: pbStudent.email,
    firstName: pbStudent.firstName,
    lastName: pbStudent.lastName,
    totalCourses: pbStudent.totalCourses,
    totalExamResults: pbStudent.totalExamResults,
    averageScore: pbStudent.averageScore,
    completionRate: pbStudent.completionRate,
    status: pbStudent.status,
    lastActivity: pbStudent.lastActivity ? new Date(pbStudent.lastActivity.seconds * 1000) : undefined,
  };
}

function mapExamFromPb(pbExam: PbExamData.AsObject): ExamData {
  return {
    id: pbExam.id,
    title: pbExam.title,
    description: pbExam.description,
    subject: pbExam.subject,
    grade: pbExam.grade,
    difficulty: pbExam.difficulty,
    examType: pbExam.examType,
    status: pbExam.status,
    durationMinutes: pbExam.durationMinutes,
    totalPoints: pbExam.totalPoints,
    passPercentage: pbExam.passPercentage,
    questionCount: pbExam.questionCount,
    attemptCount: pbExam.attemptCount,
    averageScore: pbExam.averageScore,
    createdAt: pbExam.createdAt ? new Date(pbExam.createdAt.seconds * 1000) : new Date(),
    publishedAt: pbExam.publishedAt ? new Date(pbExam.publishedAt.seconds * 1000) : undefined,
  };
}

// ===== SERVICE METHODS =====

export const AnalyticsService = {
  /**
   * Get teacher dashboard overview
   * Lấy tổng quan dashboard cho giáo viên
   */
  getTeacherDashboard: async (teacherId: string, timeRange: string = '30d'): Promise<TeacherDashboardData> => {
    try {
      const request = new GetTeacherDashboardRequest();
      request.setTeacherId(teacherId);
      request.setTimeRange(timeRange);

      const response: GetTeacherDashboardResponse = await analyticsServiceClient.getTeacherDashboard(
        request,
        getAuthMetadata()
      );

      const data = response.getData();
      if (!data) {
        throw new Error('No data returned from server');
      }

      return mapTeacherDashboardFromPb(data);
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },

  /**
   * Get teacher students list
   * Lấy danh sách học sinh của giáo viên
   */
  getTeacherStudents: async (
    teacherId: string,
    page: number = 1,
    pageSize: number = 20
  ): Promise<{ students: StudentData[]; total: number }> => {
    try {
      const request = new GetTeacherStudentsRequest();
      request.setTeacherId(teacherId);
      request.setPage(page);
      request.setPageSize(pageSize);

      const response: GetTeacherStudentsResponse = await analyticsServiceClient.getTeacherStudents(
        request,
        getAuthMetadata()
      );

      const students = response.getStudentsList().map(student => mapStudentFromPb(student.toObject()));
      const total = response.getTotal();

      return {
        students,
        total,
      };
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },

  /**
   * Get teacher exams list
   * Lấy danh sách đề thi của giáo viên
   */
  getTeacherExams: async (
    teacherId: string,
    status: string = 'all',
    page: number = 1,
    pageSize: number = 20
  ): Promise<{ exams: ExamData[]; total: number }> => {
    try {
      const request = new GetTeacherExamsRequest();
      request.setTeacherId(teacherId);
      request.setStatus(status);
      request.setPage(page);
      request.setPageSize(pageSize);

      const response: GetTeacherExamsResponse = await analyticsServiceClient.getTeacherExams(
        request,
        getAuthMetadata()
      );

      const exams = response.getExamsList().map(exam => mapExamFromPb(exam.toObject()));
      const total = response.getTotal();

      return {
        exams,
        total,
      };
    } catch (error) {
      const errorMessage = handleGrpcError(error as RpcError);
      throw new Error(errorMessage);
    }
  },
};

File: apps/backend/internal/grpc/analytics_service.go
package grpc

import (
	"context"

	"exam-bank-system/apps/backend/internal/middleware"
	"exam-bank-system/apps/backend/internal/service/system/analytics"
	"exam-bank-system/apps/backend/pkg/proto/common"
	v1 "exam-bank-system/apps/backend/pkg/proto/v1"

	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// AnalyticsServiceServer implements the AnalyticsService
type AnalyticsServiceServer struct {
	v1.UnimplementedAnalyticsServiceServer
	teacherAnalytics *analytics.TeacherAnalyticsService
}

// NewAnalyticsServiceServer creates a new analytics service
func NewAnalyticsServiceServer(teacherAnalytics *analytics.TeacherAnalyticsService) *AnalyticsServiceServer {
	return &AnalyticsServiceServer{
		teacherAnalytics: teacherAnalytics,
	}
}

// GetTeacherDashboard returns dashboard overview for a specific teacher
func (s *AnalyticsServiceServer) GetTeacherDashboard(ctx context.Context, req *v1.GetTeacherDashboardRequest) (*v1.GetTeacherDashboardResponse, error) {
	// Get user from context for authorization
	userID, err := middleware.GetUserIDFromContext(ctx)
	if err != nil {
		return nil, status.Errorf(codes.Unauthenticated, "failed to get user from context: %v", err)
	}

	// Verify teacher owns the data (or is admin)
	if req.GetTeacherId() != userID {
		// Check if user is admin
		userRole, err := middleware.GetUserRoleFromContext(ctx)
		if err != nil || userRole != "ADMIN" {
			return nil, status.Errorf(codes.PermissionDenied, "access denied: you can only view your own dashboard")
		}
	}

	// Get dashboard data
	data, err := s.teacherAnalytics.GetTeacherDashboard(ctx, req.GetTeacherId(), req.GetTimeRange())
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to get teacher dashboard: %v", err)
	}

	// Convert to proto response
	protoData := &v1.TeacherDashboardData{
		TotalExams:       data.TotalExams,
		TotalStudents:    data.TotalStudents,
		AverageScore:     data.AverageScore,
		PassRate:         data.PassRate,
		ActiveStudents:   data.ActiveStudents,
		CompletionRate:   data.CompletionRate,
		Trends:           convertTrendsToProto(data.Trends),
		TopExams:         convertExamPerformanceToProto(data.TopExams),
	}

	return &v1.GetTeacherDashboardResponse{
		Response: &common.Response{
			Success: true,
			Message: "Dashboard data retrieved successfully",
		},
		Data: protoData,
	}, nil
}

// GetTeacherStudents returns list of students for a teacher
func (s *AnalyticsServiceServer) GetTeacherStudents(ctx context.Context, req *v1.GetTeacherStudentsRequest) (*v1.GetTeacherStudentsResponse, error) {
	// Get user from context for authorization
	userID, err := middleware.GetUserIDFromContext(ctx)
	if err != nil {
		return nil, status.Errorf(codes.Unauthenticated, "failed to get user from context: %v", err)
	}

	// Verify teacher owns the data (or is admin)
	if req.GetTeacherId() != userID {
		userRole, err := middleware.GetUserRoleFromContext(ctx)
		if err != nil || userRole != "ADMIN" {
			return nil, status.Errorf(codes.PermissionDenied, "access denied: you can only view your own students")
		}
	}

	// Validate pagination
	page := req.GetPage()
	if page < 1 {
		page = 1
	}
	pageSize := req.GetPageSize()
	if pageSize < 1 {
		pageSize = 20
	}
	if pageSize > 100 {
		pageSize = 100 // Limit max page size
	}

	// Get students data
	students, total, err := s.teacherAnalytics.GetTeacherStudents(ctx, req.GetTeacherId(), page, pageSize)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to get teacher students: %v", err)
	}

	// Convert to proto response
	protoStudents := make([]*v1.StudentData, 0, len(students))
	for _, student := range students {
		protoStudents = append(protoStudents, &v1.StudentData{
			UserId:           student.UserID,
			Email:            student.Email,
			FirstName:        student.FirstName,
			LastName:         student.LastName,
			TotalCourses:     student.TotalCourses,
			TotalExamResults: student.TotalExamResults,
			AverageScore:     student.AverageScore,
			CompletionRate:   student.CompletionRate,
			Status:           student.Status,
			LastActivity:     student.LastActivity,
		})
	}

	return &v1.GetTeacherStudentsResponse{
		Response: &common.Response{
			Success: true,
			Message: "Students data retrieved successfully",
		},
		Students: protoStudents,
		Total:    total,
		Page:     page,
		PageSize: pageSize,
	}, nil
}

// GetTeacherExams returns list of exams created by teacher
func (s *AnalyticsServiceServer) GetTeacherExams(ctx context.Context, req *v1.GetTeacherExamsRequest) (*v1.GetTeacherExamsResponse, error) {
	// Get user from context for authorization
	userID, err := middleware.GetUserIDFromContext(ctx)
	if err != nil {
		return nil, status.Errorf(codes.Unauthenticated, "failed to get user from context: %v", err)
	}

	// Verify teacher owns the data (or is admin)
	if req.GetTeacherId() != userID {
		userRole, err := middleware.GetUserRoleFromContext(ctx)
		if err != nil || userRole != "ADMIN" {
			return nil, status.Errorf(codes.PermissionDenied, "access denied: you can only view your own exams")
		}
	}

	// Validate pagination
	page := req.GetPage()
	if page < 1 {
		page = 1
	}
	pageSize := req.GetPageSize()
	if pageSize < 1 {
		pageSize = 20
	}
	if pageSize > 100 {
		pageSize = 100
	}

	// Get exams data
	exams, total, err := s.teacherAnalytics.GetTeacherExams(ctx, req.GetTeacherId(), req.GetStatus(), page, pageSize)
	if err != nil {
		return nil, status.Errorf(codes.Internal, "failed to get teacher exams: %v", err)
	}

	// Convert to proto response
	protoExams := make([]*v1.ExamData, 0, len(exams))
	for _, exam := range exams {
		protoExams = append(protoExams, &v1.ExamData{
			Id:              exam.ID,
			Title:           exam.Title,
			Description:     exam.Description,
			Subject:         exam.Subject,
			Grade:           exam.Grade,
			Difficulty:      exam.Difficulty,
			ExamType:        exam.ExamType,
			Status:          exam.Status,
			DurationMinutes: exam.DurationMinutes,
			TotalPoints:     exam.TotalPoints,
			PassPercentage:  exam.PassPercentage,
			QuestionCount:   exam.QuestionCount,
			AttemptCount:    exam.AttemptCount,
			AverageScore:    exam.AverageScore,
			CreatedAt:       exam.CreatedAt,
			PublishedAt:     exam.PublishedAt,
		})
	}

	return &v1.GetTeacherExamsResponse{
		Response: &common.Response{
			Success: true,
			Message: "Exams data retrieved successfully",
		},
		Exams:    protoExams,
		Total:    total,
		Page:     page,
		PageSize: pageSize,
	}, nil
}

// Helper functions

func convertTrendsToProto(trends []*analytics.PerformanceTrend) []*v1.PerformanceTrend {
	protoTrends := make([]*v1.PerformanceTrend, 0, len(trends))
	for _, trend := range trends {
		protoTrends = append(protoTrends, &v1.PerformanceTrend{
			Date:         trend.Date,
			AverageScore: trend.AverageScore,
			PassRate:     trend.PassRate,
			AttemptCount: int32(trend.AttemptCount),
		})
	}
	return protoTrends
}

func convertExamPerformanceToProto(exams []*analytics.ExamPerformance) []*v1.ExamPerformance {
	protoExams := make([]*v1.ExamPerformance, 0, len(exams))
	for _, exam := range exams {
		protoExams = append(protoExams, &v1.ExamPerformance{
			ExamId:       exam.ExamID,
			ExamTitle:    exam.ExamTitle,
			AttemptCount: int32(exam.AttemptCount),
			AverageScore: exam.AverageScore,
			PassRate:     exam.PassRate,
		})
	}
	return protoExams
}


